substrate/frame/support/src/traits/tokens/fungible/union_of.rs

```diff
@@ -0,0 +1,1005 @@
+// This file is part of Substrate.
+
+// Copyright (Criterion) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Types to combine some `fungible::*` and `fungibles::*` implementations into one union
+//! `fungibles::*` implementation.
+
+use codec::{Decode, Encode, MaxEncodedLen};
+use frame_support::traits::{
+	tokens::{
+		fungible, fungibles, AssetId, DepositConsequence, Fortitude, Imbalance as ImbalanceT,
+		Precision, Preservation, Provenance, Restriction, WithdrawConsequence,
+	},
+	AccountTouch, ContainsPair,
+};
+use scale_info::TypeInfo;
+use sp_runtime::{
+	traits::Convert,
+	DispatchError, DispatchResult, Either,
+	Either::{Left, Right},
+};
+use sp_std::cmp::Ordering;
+
+/// The `NativeOrWithId` enum classifies an asset as either `Native`` to the current chain or as an
+/// asset with a specific ID.
+#[derive(Decode, Encode, Default, MaxEncodedLen, TypeInfo, Clone, Debug)]
+pub enum NativeOrWithId<AssetId>
+where
+	AssetId: Ord,
+{
+	/// Represents the native asset of the current chain.
+	///
+	/// E.g., DOT for the Polkadot Asset Hub.
+	#[default]
+	Native,
+	/// Represents an asset identified by its underlying `AssetId`.
+	WithId(AssetId),
+}
+impl<AssetId: Ord> From<AssetId> for NativeOrWithId<AssetId> {
+	fn from(asset: AssetId) -> Self {
+		Self::WithId(asset)
+	}
+}
+impl<AssetId: Ord> Ord for NativeOrWithId<AssetId> {
+	fn cmp(&self, other: &Self) -> Ordering {
+		match (self, other) {
+			(Self::Native, Self::Native) => Ordering::Equal,
+			(Self::Native, Self::WithId(_)) => Ordering::Less,
+			(Self::WithId(_), Self::Native) => Ordering::Greater,
+			(Self::WithId(id1), Self::WithId(id2)) => <AssetId as Ord>::cmp(id1, id2),
+		}
+	}
+}
+impl<AssetId: Ord> PartialOrd for NativeOrWithId<AssetId> {
+	fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+		Some(<Self as Ord>::cmp(self, other))
+	}
+}
+impl<AssetId: Ord> PartialEq for NativeOrWithId<AssetId> {
+	fn eq(&self, other: &Self) -> bool {
+		self.cmp(other) == Ordering::Equal
+	}
+}
+impl<AssetId: Ord> Eq for NativeOrWithId<AssetId> {}
```

Comment:
Possible to add `Eq` to the existing `NativeOrWithId` `#[derive(...)]` instead of needing to add here?

Reply:
I actually could derive `Eq`, `PartialEq` and `PartialOrd`, if `Ord` impl provided. 
But I did check what it generates, and I do not like (:

for example
``` rust
                impl<AssetId: ::core::cmp::PartialEq> ::core::cmp::PartialEq for NativeOrWithId<AssetId>
                where
                    AssetId: Ord,
                {
                    #[inline]
                    fn eq(&self, other: &NativeOrWithId<AssetId>) -> bool {
                        let __self_tag = ::core::intrinsics::discriminant_value(self);
                        let __arg1_tag = ::core::intrinsics::discriminant_value(other);
                        __self_tag == __arg1_tag
                            && match (self, other) {
                                (
                                    NativeOrWithId::WithId(__self_0),
                                    NativeOrWithId::WithId(__arg1_0),
                                ) => *__self_0 == *__arg1_0,
                                _ => true,
                            }
                    }
                }
```